lardataobj/Simulation/ParticleAncestryMap.cxx

```diff
@@ -0,0 +1,41 @@
+#ifndef PARTICLEANCESTRYMAP_CXX
+#define PARTICLEANCESTRYMAP_CXX
+
+#include "lardataobj/Simulation/ParticleAncestryMap.h"
+
+namespace sim {
+
+  void ParticleAncestryMap::SetMap(const std::map<int, std::set<int>>& map) { fParticleMap = map; }
+
+  std::map<int, std::set<int>> ParticleAncestryMap::GetMap() const { return fParticleMap; }
+
+  bool ParticleAncestryMap::HasDroppedDescendants(const int& trackid)
+  {
+    return fParticleMap.count(trackid) != 0;
+  }
+
+  std::set<int> ParticleAncestryMap::GetAllDroppedDescendants(const int& trackid)
+  {
+    return fParticleMap[trackid];
+  }
+
+  bool ParticleAncestryMap::HasAncestor(const int& trackid)
+  {
+    for (auto const& [ancestor, descendants] : fParticleMap) {
+      if (descendants.count(trackid) != 0) return true;
+    }
+
+    return false;
+  }
+
+  int ParticleAncestryMap::GetAncestor(const int& trackid)
+  {
+    for (auto const& [ancestor, descendants] : fParticleMap) {
+      if (descendants.count(trackid) != 0) return ancestor;
+    }
+
+    return -std::numeric_limits<int>::max();
```

Comment:
How do you expect users to use the value of `GetAncestor()`?  I can think of a few options:
```C++
// Option 1
if (int i = map->GetAncestor(42); i == - std::numeric_limits<int>::max()) { 
  // Ancestor doesn't exist.  Fragile due to comparison of hard-coded number
}

// Option 2
if (int i = map->GetAncestor(42); i < 0) {
  // Ancestor doesn't exist.  
  // Better due to not having to compare against hard-coded number.
}

// Option 3
if (int i = map->GetAncestor(42); !map->Exists(i)) {
  // Ancestor doesn't exist.  
  // Best option as user doesn't have to compare against any number but lets the implementation
  // decide how to do the comparison.
}
```

I suggestion using option 3, but that requires adding a new function.

Reply:
Is this not covered by the `bool HasAncestor(const int trackid)` function?

Reply:
Sorry for the delay in replying, @henrylay97.  Yes, the `HasAncestor(int trackid)` function already can return `true` or `false`.  The difference here, though, is that the code has already traversed the `std::map` when calling `map->GetAncestor(42)` and already knows that there is not an entry for 42.  The implementation of `Exists(int trackid)` would simply be something like:

```C++
bool Exists(int const trackid) const { 
  return trackid == -std::numeric_limits<int>::max(); 
}
```

This way, if you want to call `GetAncestor`, you don't have to traverse the tree again with `HasAncestor`.

Reply:
I see! I misunderstood your point I think. I will implement an `Exists` function and probably remove the existing `HasAncestor` function to avoid confusion.